python/pipeline/utils/galvo_corrections.py

```diff
@@ -313,4 +315,277 @@ def correct_motion(scan, x_shifts, y_shifts, in_place=True):
+
+    for start_idx, end_idx in tqdm(
+        zip(start_indices[:-1], start_indices[1:]), total=len(start_indices) - 1
+    ):
+
+        scan_fragment = scan[:, :, start_idx:end_idx]
+        if abs(raster_phase) > 1e-7:
+            scan_fragment = correct_raster(
+                scan_fragment, raster_phase, fill_fraction
+            )  # raster
+        scan_fragment = correct_motion(
+            scan_fragment, x_shifts[start_idx:end_idx], y_shifts[start_idx:end_idx]
+        )  # motion
+        scan[:, :, start_idx:end_idx] = scan_fragment
+
+    return scan
+
+
+def create_template(scan, key):
+    """
+    Creates the template all frames are compared against to determine the
+    amount of motion that occured. Exclusively used for the first iteration
+    of motion correction.
+    """
+
+    ## Get needed info
+    pipe = _get_pipe(key)
+    px_height, px_width, _, _ = _get_field_size_info(key)
+    skip_rows = int(
+        round(px_height * 0.10)
+    )  # we discard some rows/cols to avoid edge artifacts
+    skip_cols = int(round(px_width * 0.10))
+
+    ## Select template source
+    # Default behavior: use middle 2000 frames as template source
+    if key["motion_correction_method"] in (1,):
+        middle_frame = int(np.floor(scan.shape[-1] / 2))
+        mini_scan = scan[
+            :,
+            skip_rows:-skip_rows,
+            skip_cols:-skip_cols,
+            :,
+            max(middle_frame - 1000, 0) : middle_frame + 1000,
+        ]
+    # Use full scan as template
+    if key["motion_correction_method"] in (
+        2,
+        3,
+        4,
+        5,
+        6,
+        7,
+        8,
+    ):
+        mini_scan = scan[
+            :,
+            skip_rows:-skip_rows,
+            skip_cols:-skip_cols,
+            :,
+            :,
+        ]
+    else:
+        raise PipelineException(f"The create_template() function does not currently support motion_correction_method {key['motion_correction_method']}")
+    mini_scan = mini_scan.astype(np.float32, copy=False)
+
+    # Correct mini scan
+    correct_raster = (pipe.RasterCorrection() & key).get_correct_raster()
+    mini_scan = correct_raster(mini_scan)
+
+    # Create template
+    mini_scan = 2 * np.sqrt(mini_scan - mini_scan.min() + 3 / 8)  # *
+    template = np.mean(mini_scan, axis=-1).squeeze()
+
+    # Apply spatial filtering (if needed)
+    if key["motion_correction_method"] in (1,):
+        template = ndimage.gaussian_filter(template, 0.7)  # **
+    elif key["motion_correction_method"] in (2,3,4,5,6,7,8,):
+        pass
+    else:
+        raise PipelineException(f"The create_template() function does not currently support motion_correction_method {key['motion_correction_method']}")
+    del mini_scan
+    # * Anscombe tranform to normalize noise, increase contrast and decrease outliers' leverage
+    # ** Small amount of gaussian smoothing to get rid of high frequency noise
+
+    return template
+
+
+def create_refined_template(
```

Comment:
Consider consolidating with create_template with x_shifts/y_shifts as optional arguments with default values = None, case statements to use if non-None inputs are provided?

Reply:
After going through it, there are enough cases that are different between the two that the logic of the merged function would be confusing to follow. Future methods may require new but different processing for the first and second template, so it seems better to keep them separate for now.